### PR TITLE
methods to make fake pizza data

### PR DESCRIPTION
**Pull request by lauraannwilliams**
In addition to generating a fake pizza company (mostly drawn from Pop Culture), this will generate a variety of international toppings, combos, and styles.

I've updated:

- [X] USAGE.md docs
- [X] CHANGELOG.md

### CODE REVIEW

File: lib/faker/pizza.ex [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+  @doc """
+  Returns a vegetable string
+  """
+  @spec vegetable() :: String.t
+  sampler :vegetable, ["Artichoke Hearts", "Arugula", "Banana", "Banana Peppers", "Basil", "Black Olives", "Broccoli", "Capers", "Caramelised Onions", "Cherry Tomatoes", "Coconut", "Eggplant", "Garlic", "Green Olives", "Green Peas", "Green Peppers", "Habanero Peppers", "Jalapeños", "Mango", "Mushrooms", "Onions", "Oysters", "Peperoncini", "Pickled Ginger", "Pineapple", "Potatoes", "Red Onion", "Red Peppers", "Sauerkraut", "Spinach", "Sun-Dried Tomatoes", "Sweet Corn", "Sweet Potato", "Zucchini"]
+
+  @doc """
+  Returns a size string
+  """
+  @spec combo() :: Strint.t
```
Comment (tobyhinloopen) [inline lib/faker/pizza.ex]: I think this is a typo

File: lib/faker/pizza.ex [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+  """
+  @spec size_or_inches() :: [char]
+  def size_or_inches, do: size_or_inches(:crypto.rand_uniform(1, 3))
+  defp size_or_inches(1), do: size()
+  defp size_or_inches(2), do: inches()
+
+  @doc """
+  Returns a size string
+  """
+  @spec size() :: Strint.t
```
Comment (tobyhinloopen) [inline lib/faker/pizza.ex]: I think this is a typo

File: lib/faker/pizza.ex [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+  sampler :size, ["Personal", "Small", "Medium", "Large", "Extra-Large", "Family"]
+
+  @doc """
+  Returns a inches string
+  """
+  @spec inches() :: Strint.t
```
Comment (tobyhinloopen) [inline lib/faker/pizza.ex]: I think this is a typo

File: mix.lock [hunk 4]
```diff
@@ -1,4 +1,4 @@
...
-  "credo": {:hex, :credo, "0.7.3", "9827ab04002186af1aec014a811839a06f72aaae6cd5eed3919b248c8767dbf3", [:mix], [{:bunt, "~> 0.2.0", [hex: :bunt, repo: "hexpm", optional: false]}], "hexpm"},
-  "earmark": {:hex, :earmark, "1.2.0", "bf1ce17aea43ab62f6943b97bd6e3dc032ce45d4f787504e3adf738e54b42f3a", [:mix], [], "hexpm"},
-  "ex_doc": {:hex, :ex_doc, "0.15.1", "d5f9d588fd802152516fccfdb96d6073753f77314fcfee892b15b6724ca0d596", [:mix], [{:earmark, "~> 1.1", [hex: :earmark, repo: "hexpm", optional: false]}], "hexpm"}}
+%{"bunt": {:hex, :bunt, "0.2.0", "951c6e801e8b1d2cbe58ebbd3e616a869061ddadcc4863d0a2182541acae9a38", [:mix], []},
+  "credo": {:hex, :credo, "0.7.3", "9827ab04002186af1aec014a811839a06f72aaae6cd5eed3919b248c8767dbf3", [:mix], [{:bunt, "~> 0.2.0", [hex: :bunt, optional: false]}]},
+  "earmark": {:hex, :earmark, "1.2.0", "bf1ce17aea43ab62f6943b97bd6e3dc032ce45d4f787504e3adf738e54b42f3a", [:mix], []},
+  "ex_doc": {:hex, :ex_doc, "0.15.1", "d5f9d588fd802152516fccfdb96d6073753f77314fcfee892b15b6724ca0d596", [:mix], [{:earmark, "~> 1.1", [hex: :earmark, optional: false]}]}}
```
Comment (tobyhinloopen) [inline mix.lock]: Is this relevant to the PR? Why is this changed?

File: lib/faker/pizza.ex [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+  def toppings_and(num) do
+    num
+    |> toppings()
+    |> add_and
+  end
+
+  @doc """
+  Converts a list to a string, with "and" before the last item
+  """
+  @spec add_and(list) :: [char]
```
Comment (tobyhinloopen) [inline lib/faker/pizza.ex]: You're returning a `String.t`, right? Not a char list?

File: lib/faker/pizza.ex [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+  def toppings_and(range \\ %Range{first: 2, last: 5})
+
+  def toppings_and(%Range{first: first, last: last}) do
+    toppings_and(:crypto.rand_uniform(first, last + 1))
+  end
+
+  @doc """
+  Returns a specified number of toppings as a string, with an "and" before the last topping
+  """
+  @spec toppings_and(integer) :: [char]
```
Comment (tobyhinloopen) [inline lib/faker/pizza.ex]: You're returning a `String.t`, right? Not a char list?

**Comment by tobyhinloopen** [comment]
First of all, thank you for your PR. Who doesn't like Pizza, right?

However, and I don't know how to say this without being some kind of party pooper: A couple of "fun" based Faker modules is fine, but I don't think we want to add too many. We already have `Pokemon`, `Beer` and `Superhero`. I highly doubt anyone actually uses these, while they do add extra code to maintain.

I'm curious to know how @igas feels about adding more "fun" Faker data.

File: lib/faker/pizza.ex (hunk 2 above)
Comment (lauraannwilliams) [inline lib/faker/pizza.ex]: I got what you were referring to for most of them, but not sure what the typo is here?

File: mix.lock (hunk 4 above)
Comment (lauraannwilliams) [inline mix.lock]: I assume these were out of date on the repo. I can remove them if @igas requests :)

**Comment by lauraannwilliams** [comment]
I really appreciate you catching my various typos - I'll clean that up today. 

As for the "fun" aspect, I built this one because I had a real use case. I'm working on code to for a decision making process. We use 'Pizza' related examples all the time because they're very accessible to new users.  It lets them focus on the technology, instead of the data. 

Not that I mind 'Fun' aspects either. If you look at https://github.com/stympy/faker you'll see a lot of fun ones, that I would use - Like Star Wars, Game of Thrones, etc. If I'm populating a demo with 'fake' comments, it's nice to have usernames that make it clear it's sample data.

File: lib/faker/pizza.ex (hunk 2 above)
Comment (tobyhinloopen) [inline lib/faker/pizza.ex]: `Strint.t` is here

File: lib/faker/pizza.ex [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+  @doc """
+  Returns a vegetable string
+  """
+  @spec vegetable() :: String.t
+  sampler :vegetable, ["Artichoke Hearts", "Arugula", "Banana", "Banana Peppers", "Basil", "Black Olives", "Broccoli", "Capers", "Caramelised Onions", "Cherry Tomatoes", "Coconut", "Eggplant", "Garlic", "Green Olives", "Green Peas", "Green Peppers", "Habanero Peppers", "Jalapeños", "Mango", "Mushrooms", "Onions", "Oysters", "Peperoncini", "Pickled Ginger", "Pineapple", "Potatoes", "Red Onion", "Red Peppers", "Sauerkraut", "Spinach", "Sun-Dried Tomatoes", "Sweet Corn", "Sweet Potato", "Zucchini"]
+
+  @doc """
+  Returns a combo string
+  """
+  @spec combo() :: Strint.t
```
Comment (tobyhinloopen) [inline lib/faker/pizza.ex]: Another `Strint`

File: lib/faker/pizza.ex [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+  @doc """
+  Returns a size string
+  """
+  @spec size() :: Strint.t
+  sampler :size, ["Personal", "Small", "Medium", "Large", "Extra-Large", "Family"]
+
+  @doc """
+  Returns an inches string
+  """
+  @spec inches() :: Strint.t
```
Comment (tobyhinloopen) [inline lib/faker/pizza.ex]: Another `Strint`

**Comment by lauraannwilliams** [comment]
Thanks :)

**Comment by igas** [comment]
@tobyhinloopen I don't mind fun modules, I think it's our responsibility to make it easier to add and maintain. In our projects we use quite a lot of fun names for different demo things.

File: CHANGELOG.md [hunk 9]
```diff
@@ -9,6 +9,7 @@ Change log itself follows [Keep a CHANGELOG](http://keepachangelog.com) format.
 ## Unreleased
 
 ### Added
+* `Faker.Pizza` [@lauraannwilliams][]
```
Comment (igas) [inline CHANGELOG.md]: This is just label in markdown, you also need to add URL to your page at the end of changelog.

File: USAGE.md [hunk 10]
```diff
@@ -407,3 +408,45 @@ Faker.Superhero.name #=> "Magnificent Shatterstar"
 
 Faker.Superhero.power #=> "Photokinesis"
 ```
+
+### Faker.Pizza
+
+```elixir
+
+Faker.Pizza.pizzas(%Range{first: 2, last: 3}) => ["Medium Thai Chicken", "26\" with Eel and Salami", "Large with Squid, Duck, Meatballs, and Classic Tomato Sauce"]
```
Comment (igas) [inline USAGE.md]: we can use `2..3`

File: USAGE.md [hunk 11]
```diff
@@ -407,3 +408,45 @@ Faker.Superhero.name #=> "Magnificent Shatterstar"
...
+
+Faker.Pizza.pizzas(2) => ["30\" with Alfredo Sauce, Scallops, Lactose Free Cheese, and Cherry Tomatoes", "Family Deep Fried Pizza Double Dutch"]
```
Comment (igas) [inline USAGE.md]: please change `=>` to `# =>`

File: mix.lock (hunk 4 above)
Comment (igas) [inline mix.lock]: @lauraannwilliams I don't think so, pls update your hex by running `mix local.hex`

File: lib/faker/pizza.ex [hunk 12]
```diff
@@ -0,0 +1,162 @@
...
+    |> Enum.take(num)
+  end
+
+  @doc """
+  Returns a string with a random number of toppings (in between the specified
+  range). "And" will appear before the last topping, along with an Oxford comma.
+  If no range is specified it defaults to 2..5
+  """
+  @spec toppings_and(Range.t) :: String.t
+  def toppings_and(range \\ %Range{first: 2, last: 5})
```
Comment (igas) [inline lib/faker/pizza.ex]: maybe call this one `toppings` and current `toppings` rename to `list_of_toppings`. I'm not sure suffix `and` looks good in function name. What do you think @tobyhinloopen ?

File: lib/faker/pizza.ex [hunk 13]
```diff
@@ -0,0 +1,162 @@
...
+
+  @moduledoc """
+  Functions for generating Pizza related data in English
+  """
+
+  @doc """
+  Returns a list with a random number of pizzas (in between the specified
+  range)
+  If no range is specified it defaults to 2..5
+  """
```
Comment (igas) [inline lib/faker/pizza.ex]: If I remember correctly, this documentation will be overridden by next one because function have the same arity, it's need to be rephrased and grouped into one.

File: lib/faker/pizza.ex (hunk 12 above)
Comment (tobyhinloopen) [inline lib/faker/pizza.ex]: `toppings_sentence` combined with renaming `add_and` to `to_sentence`. Inspired by rails' `to_sentence`

File: lib/faker/pizza.ex (hunk 13 above)
Comment (lauraannwilliams) [inline lib/faker/pizza.ex]: I was emulating the style in the Lorem module. I've updated it to single documentation block

File: mix.lock (hunk 4 above)
Comment (lauraannwilliams) [inline mix.lock]: I didn't realize my hex was out of date. Thanks :)

**Comment by jessejanderson** [comment]
Some variety of "fun" modules are also nice to have when you need to generate something that there isn't a perfect match for. I needed "random event names" and ended up using some combo of the Superhero and Beer modules to get something that semi resembled an event name. ;)

**Comment by lauraannwilliams** [comment]
Please let me know if there's anything else you need from me :)

**Comment by igas** [comment]
Hi @lauraannwilliams it looks great. Only thing unrelated to your code I want to discuss with you and @tobyhinloopen. We added `to_sentence/1` which is awesome and generic, what do you think if we move it to `Faker.Util` or `Faker` to use everywhere. So we can drop `toppings_sentence/1` in this case. And use this new function directly in `pizza/1`.

**Comment by tobyhinloopen** [comment]
https://github.com/igas/faker/pull/85/files#diff-eaf63e239fc3f2d2aca36b4fb9b9ef4aR50

**Comment by lauraannwilliams** [comment]
Sounds good to me.

**Comment by igas** [comment]
I think all looks good, except `toppings_sentence` and `to_sentence`. As discussed earlier it'll be to many APIs, if we'll start adding it to every function that return list, I think it should be on consumer level through Util, and the ones that here already using it should also go through Util. So lets drop this 2 and merge it.

**Comment by lauraannwilliams** [comment]
I was waiting for the other pull request, didn't realize you were waiting on me! 

I'll yank out the to_sentence this weekend.

**Comment by igas** [comment]
Thank you @lauraannwilliams

**Comment by lauraannwilliams** [comment]
Updates complete :)

**Comment by lauraannwilliams** [comment]
I see this is failing a credo check, i'll update this week.

**Comment by igas** [comment]
@lauraannwilliams it is actually me, master have similar problem. It would be awesome if you'll find the cause. I couldn't find the time to come back to this problem. If you, me or @tobyhinloopen will not find the cause of types issue, maybe we need to turn it off, or relax versions on CI.

**Comment by igas** [comment]
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:

**Comment by lauraannwilliams** [comment]
yay!
